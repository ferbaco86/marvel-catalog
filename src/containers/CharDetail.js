import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import styled from 'styled-components';
import { fetchCharID } from '../actions/index';
import fetchCharDetail from '../api/fetCharDetail';
import LoaderSpinner from '../components/LoaderSpinner';
import CharInfo from '../components/CharInfo';
import NavBar from '../components/NavBar';
import SeriesEventsInfo from '../components/SeriesEventsInfo';
import HeroCard from '../components/HeroCard';

const Container = styled.div`
display: flex;
flex-direction: column;
`;

const SeriesEventsContainer = styled.div`
display: flex;
flex-wrap: wrap;
justify-content: center;
padding: 0 6.5rem;`;

const CharDetail = props => {
  const { match } = props;
  const { detail } = useSelector(state => state);
  const dispatch = useDispatch();

  const getCharID = () => {
    dispatch(fetchCharID(match.params.id));
  };

  useEffect(() => {
    getCharID();
    dispatch(fetchCharDetail());
  }, []);

  const shouldComponentRender = () => {
    let isPending = false;
    if (detail.pending === false || detail.error !== null) {
      isPending = false;
    } else {
      isPending = true;
    }
    return isPending;
  };

  if (shouldComponentRender()) return <LoaderSpinner />;

  return (
    <>
      <NavBar />
      <Container>
        { detail.id && (detail.detail.map(char => (
          <>
            <CharInfo
              key={char.id}
              id={char.id}
              name={char.name}
              description={char.description}
              image={char.thumbnail.path}
              extension={char.thumbnail.extension}
              url={char.urls[0].url}
            />
            <SeriesEventsInfo title="Latest Series" />
            <SeriesEventsContainer>
              { detail.series.map(serie => (
                <HeroCard
                  key={serie.id}
                  name={serie.title}
                  image={serie.thumbnail.path}
                  extension={serie.thumbnail.extension}
                  url={serie.urls[0].url}
                />
              ))}
            </SeriesEventsContainer>
            <SeriesEventsInfo title="Latest Events" />
            <SeriesEventsContainer>
              { detail.events.map(event => (
                <HeroCard
                  key={event.id}
                  name={event.title}
                  image={event.thumbnail.path}
                  extension={event.thumbnail.extension}
                  url={event.urls[0].url}
                />
              ))}
            </SeriesEventsContainer>
          </>
        )))}
      </Container>
    </>
  );
};
CharDetail.propTypes = {
  match: PropTypes.number.isRequired,
};

export default CharDetail;
