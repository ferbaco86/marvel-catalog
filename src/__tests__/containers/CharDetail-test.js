import { render, screen, act } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../reducers/index';
import CharDetail from '../../containers/CharDetail';

describe('Rendering component', () => {
  it('creates a CharDetail component', () => {
    act(() => {
      render(
        <BrowserRouter>
          <Provider store={store}>
            <CharDetail match={{ params: { id: '2' } }} />
          </Provider>
        </BrowserRouter>,
      );
    });
    screen.getByText('LOADING');
  });
});

describe('Display', () => {
  it('renders correctly', () => {
    const snap = renderer.create(
      <BrowserRouter>
        <Provider store={store}>
          <CharDetail match={{ params: { id: '2' } }}/>
        </Provider>
      </BrowserRouter>,
    ).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
