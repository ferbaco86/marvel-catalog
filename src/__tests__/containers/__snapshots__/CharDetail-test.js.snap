// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display renders correctly 1`] = `
<div
  className="sc-bdfBwQ jaWpEd"
>
  LOADING
  <div
    className="sc-gsTCUz hDkRRh"
  />
</div>
`;
